tools/storage.py

```diff
@@ -41,29 +43,144 @@ def __init__(self, batch):
+        return "{}:{}".format(self.key.decode("utf-8"),
+                              self.value.decode("utf-8"))
+
+    @staticmethod
+    def record_from_bytes(b):
+        (size, v) = read_varlong(b, 0)
+
+        record_attr = b[v]
+        v = v + 1
+
+        (timestamp_delta, v) = read_varlong(b, v)
+        (offset_delta, v) = read_varlong(b, v)
+
+        (key_length, v) = read_varlong(b, v)
+        key = b[v:v + key_length]
+        v = v + key_length
+
+        (value_length, v) = read_varlong(b, v)
+        value = b[v:v + value_length]
+        v = v + value_length
+
+        (header_count, v) = read_varlong(b, v)
+        headers = []
+        for i in range(0, header_count):
+            (header_length, v) = read_varlong(b, v)
+            header = b[v:v + header_length]
+            v = v + header_length
+            headers.append(header)
+
+        return Record(size, record_attr, timestamp_delta, offset_delta, key,
+                      headers, value)
+
+    @staticmethod
+    def from_bytes(b, count):
+        records = []
+        for i in range(0, count):
+            record = Record.record_from_bytes(b)
+            b = b[:record.size]
+            records.append(record)
+        return records
+
+
 class Batch:
     def __init__(self, index, header, records):
         self.index = index
         self.header = header
+        self.records = records
 
+    def verify_crc(self):
         header_crc_bytes = struct.pack(
             "<" + HDR_FMT_RP_PREFIX_NO_CRC + HDR_FMT_CRC, *self.header[1:])
-        header_crc = crc32c.crc32(header_crc_bytes)
+        header_crc = crc32c.crc32c(header_crc_bytes)
         if self.header.header_crc != header_crc:
             raise CorruptBatchError(self)
 
-        crc = crc32c.crc32(self._crc_header_be_bytes())
-        crc = crc32c.crc32(records, crc)
+        crc = crc32c.crc32c(self._crc_header_be_bytes())
+        crc = crc32c.crc32c(self.records, crc)
         if self.header.crc != crc:
             raise CorruptBatchError(self)
 
     def last_offset(self):
-        return self.header.base_offset + self.header.record_count - 1
+        return self.header.base_offset + self.header.delta
 
     def _crc_header_be_bytes(self):
         # encode header back to big-endian for crc calculation
         return struct.pack(">" + HDR_FMT_CRC, *self.header[5:])
 
+    def parse_records(self):
+        batch = self.records
+        try:
+            if self.header.attrs == 0:
+                pass
+            elif self.header.attrs == 1:
+                import gzip
+                logger.warning("Found zip")
+                batch = gzip.decompress(self.records)
+            elif self.header.attrs == 2:
+                "snappy"
+                raise NotImplementedError(
+                    "Unsupported compression type snappy. Patches welcome!")
+            elif self.header.attrs == 3:
+                import lz4.frame
+                logger.warning("Found lz4")
```

Comment:
logger.trace? i think warning means something bad no?

Reply:
Those shouldn't have been committed, I was trying to see if they would fire on internal topics or something to see if the untested lz4/gzip code actually worked. Good catch.